src/test/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitCheckerTest.java

```diff
@@ -0,0 +1,181 @@
+                                        .body(response.getBodyAsString()
+                                                .replace("https://api.github.com/",
+                                                        "http://localhost:" + githubApi.port() + "/")
+                                                .replace("https://raw.githubusercontent.com/",
+                                                        "http://localhost:" + githubRaw.port() + "/")
+                                        )
+                                        .build();
+                            }
+                            return response;
+                        }
+
+                        @Override
+                        public String getName() {
+                            return "url-rewrite";
+                        }
+
+                    })
+    );
+    private GitHub github;
+
+    @Before
+    public void prepareMockGitHub() throws Exception {
+        new File("src/test/resources/api/mappings").mkdirs();
+        new File("src/test/resources/api/__files").mkdirs();
+        new File("src/test/resources/raw/mappings").mkdirs();
+        new File("src/test/resources/raw/__files").mkdirs();
+        githubApi.enableRecordMappings(new SingleRootFileSource("src/test/resources/api/mappings"),
+                new SingleRootFileSource("src/test/resources/api/__files"));
+        githubRaw.enableRecordMappings(new SingleRootFileSource("src/test/resources/raw/mappings"),
+                new SingleRootFileSource("src/test/resources/raw/__files"));
+
+
+        githubApi.stubFor(
+                get(urlMatching(".*")).atPriority(10).willReturn(aResponse().proxiedFrom("https://api.github.com/")));
+        githubRaw.stubFor(get(urlMatching(".*")).atPriority(10)
+                .willReturn(aResponse().proxiedFrom("https://raw.githubusercontent.com/")));
+
+        github = Connector.connect("http://localhost:" + githubApi.port(), null);
+
+        resetAllScenarios();
+    }
+
+    @Test(timeout=120000)
+    public void ThrottleOnOverTest() throws Exception {
+        githubApi.stubFor(
+                get(urlEqualTo("/rate_limit"))
+                        .inScenario("API Rate Limit Throttle on Over")
+                        .whenScenarioStateIs(Scenario.STARTED)
+                        .willReturn(
+                                aResponse()
+                                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                                        .withHeader("Accept-Encoding", "gzip,deflate")
+                                        .withBody("{ \"rate\": { \"limit\": 100, \"remaining\": 8, \"reset\": 1 } }"))
+                        .willSetStateTo("API Rate Limit Depleted"));
+
+        githubApi.stubFor(
+                get(urlEqualTo("/rate_limit"))
+                        .inScenario("API Rate Limit Throttle on Over")
+                        .whenScenarioStateIs("API Rate Limit Depleted")
```

Comment:
I like the meaningful state names.  With the complexity of the scenarios I talked about, you might also need to provide a comment above the stubs to explain what they cover.